cabal-validate/src/Main.hs

```diff
@@ -0,0 +1,430 @@
+              ++ jobsArgs opts
+              ++ extraArgs
+          )
+
+      runCabalTests suite = runCabalTests' suite []
+
+  runCabalTests' "unit-tests" ["--with-ghc=" <> compilerExecutable (compiler opts)]
+  runCabalTests "check-tests"
+  runCabalTests "parser-tests"
+  runCabalTests "rpmvercmp"
+  runCabalTests "no-thunks-test"
+
+  runHackageTests opts
+
+-- | Run Hackage tests, if enabled.
+runHackageTests :: Opts -> IO ()
+runHackageTests opts
+  | NoHackageTests <- hackageTests opts = pure ()
+  | otherwise = do
+      command <- cabalListBin opts "Cabal-tests:test:hackage-tests"
+
+      let
+        -- See #10284 for why this value is pinned.
+        hackageTestsIndexState = "--index-state=2024-08-25"
+
+        hackageTest args =
+          timedWithCwd
+            opts
+            "Cabal-tests"
+            command
+            (args ++ [hackageTestsIndexState])
+
+      hackageTest ["read-fields"]
+
+      case hackageTests opts of
+        CompleteHackageTests -> do
+          hackageTest ["parsec"]
+          hackageTest ["roundtrip"]
+        PartialHackageTests -> do
+          hackageTest ["parsec", "d"]
+          hackageTest ["roundtrip", "k"]
+
+-- | Run @cabal-testsuite@ with the @Cabal@ library with a non-default GHC.
+libSuiteWith :: Opts -> FilePath -> [String] -> IO ()
+libSuiteWith opts ghc extraArgs =
+  timedCabalBin
+    opts
+    "cabal-testsuite"
+    "exe:cabal-tests"
+    ( [ "--builddir=" <> cabalTestsuiteBuildDir opts
+      , "--with-ghc=" <> ghc
+      , -- This test suite doesn't support `--jobs` _or_ `--num-threads`!
+        "-j" <> show (jobs opts)
+      ]
+        ++ tastyArgs opts
+        ++ extraArgs
+    )
+
+-- | Run @cabal-testsuite@ with the @Cabal@ library with the default GHC.
+libSuite :: Opts -> IO ()
+libSuite opts = libSuiteWith opts (compilerExecutable (compiler opts)) (rtsArgs opts)
+
+-- | Run @cabal-testsuite@ with the @Cabal@ library with all extra GHCs.
+libSuiteExtras :: Opts -> IO ()
+libSuiteExtras opts = forM_ (extraCompilers opts) $ \compiler' ->
+  libSuiteWith opts compiler' []
+
+-- | Test the @cabal-install@ executable.
+--
+-- These tests mostly run sequentially, so they're pretty slow as a result.
+cliTests :: Opts -> IO ()
+cliTests opts = do
+  -- These are sorted in asc time used, quicker tests first.
+  timedCabalBin
+    opts
+    "cabal-install"
+    "test:long-tests"
+    ( jobsArgs opts
+        ++ tastyArgs opts
+    )
+
+  -- This doesn't work in parallel either.
+  timedCabalBin
+    opts
+    "cabal-install"
+    "test:unit-tests"
+    ( ["--num-threads", "1"]
+        ++ tastyArgs opts
+    )
+
+  -- Only single job, otherwise we fail with "Heap exhausted"
+  timedCabalBin
+    opts
+    "cabal-install"
+    "test:mem-use-tests"
+    ( ["--num-threads", "1"]
+        ++ tastyArgs opts
+    )
+
+  -- This test-suite doesn't like concurrency
+  timedCabalBin
+    opts
+    "cabal-install"
+    "test:integration-tests2"
+    ( [ "--num-threads"
```

Comment:
Could replace the ++ with multiple :-applications. I think we have some hlint integration but it probably has a lot of stuff turned off because we didn’t want to update the whole codebase at once. May be good to research how to make new packages like cabal-validate get covered by hlint without (with minimal) exceptions.

Reply:
I think that would reduce readability for negligible gain.